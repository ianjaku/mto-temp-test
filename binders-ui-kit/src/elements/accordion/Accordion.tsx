import * as React from "react";
import Icon from "../icons";
import autobind from "class-autobind";
import classnames from "classnames";
import "./accordion.styl";

export interface IAccordionProps {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    header: React.ReactElement<any> | string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    children?: React.ReactElement<any> | Array<React.ReactElement<any>>;
    onToggle?: () => void;
    onOpen?: () => void;
    isOpened?: boolean;
    className?: string;
    noGaps?: boolean;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
class Accordion extends React.Component<IAccordionProps, any> {

    private static defaultProps = {
        isOpened: false,
    };

    constructor(props: IAccordionProps) {
        super(props);
        autobind(this, Accordion.prototype);
    }

    public renderHeader(): JSX.Element {
        const { header, isOpened } = this.props;
        const openedClass = { "is-opened": isOpened };
        const arrowClasses = classnames("accordion-arrow", openedClass);
        const headerClasses = classnames("accordion-header", openedClass);

        return (
            <div className={headerClasses} onClick={this.onToggle}>
                <Icon name="arrow_drop_down" className={arrowClasses} />
                {header}
            </div>
        );
    }

    public renderChildren(): JSX.Element {
        const { children, isOpened } = this.props;

        return isOpened && (
            <div className="accordion-body">
                {children}
            </div>
        );
    }

    public render(): JSX.Element {
        const { isOpened, noGaps } = this.props;
        return (
            <div className={classnames("accordion", this.props.className, { "is-opened": isOpened }, { "accordion--noGaps": noGaps })}>
                {this.renderHeader()}
                {this.renderChildren()}
            </div>
        );
    }

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    private onToggle(e) {
        const { isOpened, onOpen, onToggle } = this.props;
        if (onToggle) {
            onToggle();
        }
        if (onOpen && !isOpened) {
            onOpen();
        }
    }
}

export default Accordion;
