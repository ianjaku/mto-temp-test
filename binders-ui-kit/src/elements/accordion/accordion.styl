@require "../../vars.styl";
 
.accordion
    position: relative
    outline: solid 1px borderGrayColor
    background-color: whiteColor
    .accordion-header
        cursor: pointer
        display: flex
        height: 48px
        align-items: center
        font-family: defaultFontName
        font-size: 14px
        .accordion-arrow
            transform: rotate(-90deg)
            transition: all .2s linear
    &.is-opened&:not(.accordion--noGaps)
        margin: 16px 0
        .accordion-header
            border-bottom: solid 1px borderGrayColor
            .accordion-arrow
                transform: rotate(0deg)
    .accordion-body
        padding: 12px
        font-family: defaultFontName
        font-size: 12px
