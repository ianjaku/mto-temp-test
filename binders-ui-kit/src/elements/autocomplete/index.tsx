import * as React from "react";
import Chip from "@material-ui/core/Chip";
import { ChipComponentProps } from "./contract";
import CircularProgress from "../circularprogress";
import ClearIcon from "@material-ui/icons/Clear";
import { TranslationKeys as TK } from "@binders/client/lib/react/i18n/translations";
import Tooltip from "@material-ui/core/Tooltip";
import autobind from "class-autobind";
import colors from "../../variables";
import cx from "classnames";
import { withTranslation } from "@binders/client/lib/react/i18n";
import "./autocomplete.styl";

export interface IAutocompleteItem {
    label: string;
    rawLabel: string;
    id: string;
    value: string;
    isNew?: boolean;
}

enum LoadingState {
    Loading = "Loading",
    Initial = "Initial",
    Complete = "Complete",
}

export interface IAutocompleteProps {
    data: IAutocompleteItem[];
    onUpdateSearchTerm?: (term: string) => void;
    // eslint-disable-next-line @typescript-eslint/ban-types
    onAddNewClick: (item: object) => void;
    // eslint-disable-next-line @typescript-eslint/ban-types
    onDeleteChipClick: (item: object) => void;
    selectedItems: IAutocompleteItem[];
    totalResults?: number;
    maxAllowedSelectedItems?: number;
    ChipComponent?: React.FC<ChipComponentProps> | React.ComponentClass<ChipComponentProps>;
    wrapChips?: boolean;
    saveOnBlur?: boolean;
    placeholder?: string;
    onInputRef?: (ref: HTMLTextAreaElement) => void;
    isLoading?: boolean;
    renderLeadingSlot?: () => React.ReactNode;
    renderTrailingSlot?: () => React.ReactNode;
    allowItemCreation?: boolean;
    noMatchesOverride?: string;
    hideIcon?: boolean;
}

export interface IAutocompleteState {
    searchTerm: string;
    loadingState?: LoadingState;
}

class Autocomplete extends React.Component<IAutocompleteProps, IAutocompleteState> {

    private autocomplete = null;
    private t;

    constructor(props) {
        super(props);
        this.t = props.t;
        autobind(this);
        this.state = {
            searchTerm: "",
            loadingState: LoadingState.Initial,
        };
    }

    public componentDidUpdate(prevProps: IAutocompleteProps) {
        const { isLoading: wasLoading } = prevProps;
        if (!wasLoading && this.props.isLoading) {
            this.setState({
                loadingState: LoadingState.Loading,
            });
        }
        if (wasLoading && !this.props.isLoading) {
            this.setState({
                loadingState: LoadingState.Complete,
            });
        }
    }

    private maybeAddMoreArea() {
        const { selectedItems, maxAllowedSelectedItems, placeholder, onInputRef } = this.props;

        if (selectedItems.length >= maxAllowedSelectedItems) {
            return null;
        }

        const isClean = selectedItems.length === 0;
        const plHolder = isClean ?
            (placeholder || this.t(TK.User_AddUserOrUsergroup)) :
            this.t(TK.General_AddMore);
        return (
            <textarea
                rows={1}
                placeholder={plHolder}
                onChange={this.onChange}
                onKeyPress={this.onRequestAdd}
                onBlur={this.onBlur}
                className="autocomplete-textarea"
                value={this.state.searchTerm}
                ref={onInputRef}
            />
        );
    }

    public render() {
        const { selectedItems, wrapChips, renderLeadingSlot, renderTrailingSlot, hideIcon } = this.props;
        const isClean = selectedItems.length === 0;
        const className = cx("autocomplete", {
            "autocomplete--is-dirty": !isClean,
            "autocomplete--hideIcon": hideIcon || !!renderLeadingSlot,
            "autocomplete--loadingComplete": this.state.loadingState === LoadingState.Complete,
        });
        return (
            <div className="autocomplete-wrapper">
                <div
                    ref={autocomplete => { this.autocomplete = autocomplete; }}
                    className={className}
                >
                    {renderLeadingSlot && renderLeadingSlot()}
                    <div className="autocomplete-pane" style={wrapChips ? { flexFlow: "wrap" } : {}}>
                        <div className="autocomplete-pane-chips">
                            {this.renderChips()}
                            {this.maybeAddMoreArea()}
                        </div>
                    </div>
                    {renderTrailingSlot && renderTrailingSlot()}
                    {this.props.isLoading && (
                        <div className="autocomplete-alignEnd">
                            {CircularProgress("", {}, undefined, colors.disabledColor)}
                        </div>
                    )}
                </div>
                {this.renderPrompts()}
            </div>
        );
    }

    private onChange({ target: { value } }) {
        this.setState({
            searchTerm: value,
            loadingState: LoadingState.Initial,
        });
        if (this.props.onUpdateSearchTerm) {
            this.props.onUpdateSearchTerm(value);
        }
    }

    /* RENDERS */

    private renderChip(label: string, value: string, isNew: boolean, isUser: boolean) {
        const { ChipComponent, maxAllowedSelectedItems } = this.props;
        const usesCustomChip = !!ChipComponent;
        const ChipToRender = this.props.ChipComponent || Chip;
        return (
            <div className={cx("autocomplete-chipWrapper", { "autocomplete-chipWrapper--singleChip": maxAllowedSelectedItems === 1 })}>
                <ChipToRender
                    classes={{
                        root: `autocomplete-chip ${isUser ? "autocomplete-chip--user" : "autocomplete-chip--group"}`,
                    }}
                    className={"autocomplete-chip"}
                    deleteIcon={<ClearIcon style={{ fill: colors.disabledColor, marginTop: "1px", width: "14px" }} />}
                    onDelete={this.onDeleteChip(value, label, isNew)}
                    label={label}
                    value={value}
                    {...(usesCustomChip ? { isNew } : {})}
                />
            </div>
        );
    }

    private renderChips() {
        const { selectedItems } = this.props;

        return selectedItems.map(({ id, label, value, isNew = false }) => {
            const tooltipProps = isNew ?
                { title: this.t(TK.User_NoUserOrUserGroup) } :
                { title: "", open: false }
            return (
                <Tooltip key={value} {...tooltipProps}>
                    <>
                        {this.renderChip(label, value, isNew, id.startsWith("uid-"))}
                    </>
                </Tooltip>
            );
        });
    }

    private renderPrompts() {
        const { totalResults, noMatchesOverride } = this.props;
        const items = this.getFilteredData();
        if (this.state.loadingState === LoadingState.Complete && items.length === 0) {
            return (
                <ul
                    className={cx("autocomplete-prompt", "autocomplete-prompt--infoMsg")}
                    style={{
                        width: this.autocomplete ? this.autocomplete.clientWidth : "auto",
                    }}
                >
                    <li className="autocomplete-prompt-item">
                        {noMatchesOverride ?? this.t(TK.General_NoMatches)}
                    </li>
                </ul>
            )
        }
        const onAutocompleteSelection = (e: React.MouseEvent, { label, rawLabel, value, id }: IAutocompleteItem) => {
            e.stopPropagation();
            this.addItem({ label, rawLabel, value, id, isNew: false });
        }
        const hasMoreResults = totalResults ? totalResults >= 10 && items.length < totalResults : false;
        return items.length === 0 ?
            <div /> :
            (
                <ul className="autocomplete-prompt">
                    {items.map(item => (
                        <li
                            key={item.value}
                            className={"autocomplete-prompt-item"}
                            onClick={e => onAutocompleteSelection(e, item)}
                        >
                            {item.label}
                        </li>
                    ))}
                    {hasMoreResults && (
                        <li key="more-results" className={"autocomplete-prompt-item--more-results"}>
                            {this.t(TK.General_MoreResultsRefineQuery, { count: totalResults - items.length })}
                        </li>
                    )}
                </ul>
            );
    }

    /* HELPER METHODS */
    private onRequestAdd(e) {
        if (e.key.toUpperCase() === "ENTER") {
            e.preventDefault();
            this.onEnter(e);
        }
    }

    private addItem({ label, rawLabel, value, id, isNew = false }) {
        this.props.onAddNewClick({ label, rawLabel, value, id, isNew });
        this.setState({
            searchTerm: "",
            loadingState: LoadingState.Initial,
        });
        setTimeout(() => this.autocomplete.scrollTop = this.autocomplete.scrollHeight, 100);
    }

    /* EVENT HANDLERS */

    private onDeleteChip(value, label, isNew) {
        return () => this.props.onDeleteChipClick({ isNew, label, value });
    }

    private onEnter(e) {
        // todo scroll to bottom
        const searchTerm = e.target.value;
        if (searchTerm) {
            const proposals = this.getFilteredData();
            if (proposals.length) {
                this.addItem(proposals[0]);
                return;
            }
            if (this.props.allowItemCreation) {
                this.addItem(this.getEmptyPossibleAnswer(searchTerm));
            }
        }
    }

    private onBlur(e) {
        if (this.props.saveOnBlur) {
            const filteredItems = this.getFilteredData();
            if (!filteredItems.length) { // only save on blur if there is no match
                this.onEnter(e);
            }
        }
    }

    private getEmptyPossibleAnswer(searchTerm: string): IAutocompleteItem {
        return {
            id: "",
            isNew: true,
            label: searchTerm,
            rawLabel: searchTerm,
            value: searchTerm,
        };
    }

    private getFilteredData(): IAutocompleteItem[] {
        const { searchTerm } = this.state;
        if (searchTerm.length < 3) {
            return [];
        }
        const { data, selectedItems } = this.props;
        return data
            .filter(({ value }) => ( // filter out items that are already selected
                !selectedItems.some(({ value: selectedValue }) => selectedValue === value)
            ))
            .filter(item => ( // filter out items that don't match search term
                item.label.toLowerCase().indexOf(searchTerm.toLowerCase()) >= 0
            ));
    }
}

export default withTranslation()(Autocomplete);
