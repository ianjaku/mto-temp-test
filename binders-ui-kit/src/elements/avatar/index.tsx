import * as React from "react";
import Avatar from "@material-ui/core/Avatar";

export interface IAvatarProps {
    image: string;
    size?: number;
    // eslint-disable-next-line @typescript-eslint/ban-types
    style?: object;
}

const avatar: React.StatelessComponent<IAvatarProps> = props => (
    <Avatar
        src={props.image}
        style={{ ...props.style, width: `${props.size}px`, height: `${props.size}px` }}
    />
);

export default avatar;





