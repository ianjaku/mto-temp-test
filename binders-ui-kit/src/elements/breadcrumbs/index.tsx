import * as React from "react";
import BreadcrumbItem from "./item";
import CaretRight from "../icons/CaretRight";
import { Link } from "react-router-dom";
import debounce from "lodash.debounce";
import "./breadcrumbs.styl";

export interface IBreadcrumbItem {
    name: string;
    link?: string;
    readonly?: boolean;
    strikeThrough?: boolean;
    tooltip?: string;
    renderAsLast?: boolean;
}

export interface IBreadcrumbsProps {
    items: IBreadcrumbItem[];
    onClick?: (object) => void;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    itemContextMenu?: ((data) => React.ReactElement<any>) | JSX.Element;
    keySuffix?: string;
}

export interface IBreadcrumbState {
    breadcrumbsEl?: HTMLElement;
    itemsWrapperEl?: HTMLElement;
    hiddenElementsCount: number;
    truncatedElementsCount: number;
    lastClipOperation?: ClipOperation;
}

enum ClipOperation {
    Truncate,
    Hide,
}

class Breadcrumbs extends React.Component<IBreadcrumbsProps, IBreadcrumbState> {
    private onResize: () => void;

    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    constructor(props) {
        super(props);
        this.onResize = debounce(this.reMaybeClipBreadcrumbs.bind(this), 500);
        window.addEventListener("resize", this.onResize);
        this.maybeClipBreadcrumbs = this.maybeClipBreadcrumbs.bind(this);
        this.setItemsWrapperRef = this.setItemsWrapperRef.bind(this);
        this.setBreadcrumbsElRef = this.setBreadcrumbsElRef.bind(this);
        this.state = {
            hiddenElementsCount: 0,
            truncatedElementsCount: 0,
        };
    }

    public componentWillUnmount(): void {
        window.removeEventListener("resize", this.onResize);
    }

    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    public componentDidUpdate(_prevProps, prevState): void {
        const { itemsWrapper: prevItemsWrapper, breadcrumbsEl: prevBreadcrumbsEl } = prevState;
        const { itemsWrapperEl, breadcrumbsEl } = this.state;
        const breadcrumbsElChanged = breadcrumbsEl !== prevBreadcrumbsEl;
        const itemsWrapperElChanged = itemsWrapperEl !== prevItemsWrapper;
        if (breadcrumbsElChanged || itemsWrapperElChanged) {
            this.maybeClipBreadcrumbs();
        }
    }

    private maybeClipBreadcrumbs() {
        const { hiddenElementsCount, truncatedElementsCount, itemsWrapperEl, breadcrumbsEl, lastClipOperation } = this.state;
        let stateUpdate;
        let newLastClipOperation;
        if (lastClipOperation === undefined || (lastClipOperation === ClipOperation.Hide)) {
            stateUpdate = {
                truncatedElementsCount: truncatedElementsCount + 1,
            }
            newLastClipOperation = ClipOperation.Truncate;
        } else {
            stateUpdate = {
                hiddenElementsCount: hiddenElementsCount + 1,
            }
            newLastClipOperation = ClipOperation.Hide;
        }
        if (itemsWrapperEl.clientWidth > breadcrumbsEl.clientWidth) {
            this.setState({
                lastClipOperation: newLastClipOperation,
                ...stateUpdate,
            });
        }
    }

    private reMaybeClipBreadcrumbs() {
        this.setState({
            hiddenElementsCount: 0,
            truncatedElementsCount: 0,
        }, () => this.maybeClipBreadcrumbs());
    }

    private setItemsWrapperRef(ref) {
        const { itemsWrapperEl } = this.state;
        if (itemsWrapperEl === undefined) {
            this.setState({ itemsWrapperEl: ref });
        }
    }

    private setBreadcrumbsElRef(ref) {
        const { breadcrumbsEl } = this.state;
        if (breadcrumbsEl === undefined) {
            this.setState({ breadcrumbsEl: ref });
        }
    }

    public render(): JSX.Element {
        return (
            <div className="breadcrumbs" ref={this.setBreadcrumbsElRef}>
                <div className="breadcrumbs-itemswrapper" ref={this.setItemsWrapperRef}>
                    {this.renderItems()}
                </div>
            </div>
        );
    }

    private renderItems() {
        const { items, onClick, itemContextMenu, keySuffix } = this.props;

        const { hiddenElementsCount, truncatedElementsCount } = this.state;

        const generalMaxWidth = this.calculateMaxBreadcrumbWidth();
        const hasItemContextMenu = !!itemContextMenu;

        const isFirstLinkEmpty = (items.at(0)?.name?.length ?? 0) === 0;

        return items
            .slice(hiddenElementsCount + (isFirstLinkEmpty ? 1 : 0))
            .reduce((acc, { name, link, readonly, strikeThrough, tooltip, renderAsLast }, index, arr) => {
                const isLast = renderAsLast !== undefined ? renderAsLast : (arr.length === index + 1);
                const previousLink = index > 0 && items[index - 1].link;
                const isOffset = hiddenElementsCount > 0;
                const isEllipsis = isOffset && index === 0;

                const maxWidth = index < truncatedElementsCount ? generalMaxWidth : undefined;

                const breadCrumbItem = (
                    <BreadcrumbItem
                        key={`item-${index}${keySuffix}`}
                        onClick={onClick}
                        name={isEllipsis ? "..." : name}
                        link={link}
                        isLast={isLast}
                        isReadonly={readonly}
                        isStrikeThrough={strikeThrough}
                        tooltip={tooltip}
                        maxWidth={maxWidth}
                        itemContextMenu={itemContextMenu}
                        hasItemContextMenu={hasItemContextMenu}
                    />
                );

                const arrow = (previousLink || link ?
                    (
                        <Link
                            className="breadcrumbs-arrow"
                            key={`carrret-${index}${keySuffix}`}
                            to={previousLink || link}
                            tabIndex={-1}
                        >
                            {{ ...CaretRight("") }}
                        </Link>
                    ) :
                    (
                        <div className="breadcrumbs-arrow">
                            {{ ...CaretRight("", {}, `carrret-${index}${keySuffix}`) }}
                        </div>
                    )
                );

                return [
                    ...acc,
                    index > 0 ? arrow : null,
                    breadCrumbItem,
                ];
            }, []);
    }

    private calculateMaxBreadcrumbWidth(): number {
        const { items } = this.props;
        const { breadcrumbsEl } = this.state;
        const { hiddenElementsCount } = this.state;
        const visibleItemsCount = items.length - hiddenElementsCount;
        return breadcrumbsEl && (breadcrumbsEl.clientWidth / visibleItemsCount) || undefined;
    }
}

export default Breadcrumbs;
