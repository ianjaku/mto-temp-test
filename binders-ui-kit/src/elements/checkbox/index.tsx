import * as React from "react";
import CheckBoxIcon from "@material-ui/icons/CheckBox";
import CheckBoxOutlineBlankIcon from "@material-ui/icons/CheckBoxOutlineBlank";
import FormControlLabel from "@material-ui/core/FormControlLabel";
import MaterialCheckbox from "@material-ui/core/Checkbox";
import Tooltip from "@material-ui/core/Tooltip";

export interface ICheckboxState {
    checked: boolean;
}

class Checkbox extends React.Component<ICheckboxProps, ICheckboxState> {
    public static defaultProps: Partial<ICheckboxProps> = {
        checked: false,
        disabled: false,
        isControlled: false,
    };

    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    static getDerivedStateFromProps(nextProps, prevState) {
        const { checked } = prevState;
        const { isControlled } = nextProps;
        return isControlled && checked !== nextProps.checked ?
            { checked: nextProps.checked } :
            null;
    }

    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    constructor(props) {
        super(props);
        this.onChange = this.onChange.bind(this);
        this.renderMaterialCheckbox = this.renderMaterialCheckbox.bind(this);
        this.state = {
            checked: props.checked,
        };
    }

    public render(): JSX.Element {
        const { label, labelStyle, tooltip } = this.props;
        const component = !label ?
            this.renderMaterialCheckbox() :
            (
                <FormControlLabel
                    label={label}
                    control={this.renderMaterialCheckbox()}
                    style={labelStyle}
                    className="checkboxControlLabel"
                />
            );
        return tooltip ? <Tooltip title={tooltip} placement="bottom-start"><div>{component}</div></Tooltip> : component;
    }

    private renderMaterialCheckbox() {
        const { disabled, iconStyle, className, iconSize } = this.props;

        const { checked } = this.state;
        return (
            <MaterialCheckbox
                checked={checked}
                disabled={disabled}
                onChange={this.onChange}
                className={className}
                icon={<CheckBoxOutlineBlankIcon fontSize={iconSize || "small"} />}
                checkedIcon={<CheckBoxIcon style={iconStyle} fontSize={iconSize || "small"} />}
                color="primary"
            />
        );
    }

    private onChange(e) {
        const { onCheck } = this.props;
        const { checked } = e.target;

        this.setState({ checked }, () => {
            if (typeof onCheck === "function") {
                onCheck(checked);
            }
        });
    }

    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    componentDidUpdate(prevProps): void {
        if (prevProps.checked !== this.props.checked) {
            this.setState({
                checked: this.props.checked
            });
        }
    }

}

export interface ICheckboxProps {
    onChange?: (isChecked: boolean) => void;
    onCheck?: (isChecked: boolean) => void;
    label?: string;
    disabled?: boolean;
    checked?: boolean;
    // eslint-disable-next-line @typescript-eslint/ban-types
    iconStyle?: object;
    isControlled?: boolean;
    // eslint-disable-next-line @typescript-eslint/ban-types
    labelStyle?: object;
    onFocus?: (e) => void;
    // eslint-disable-next-line @typescript-eslint/ban-types
    style?: object;
    className?: string;
    tooltip?: string;
    iconSize?: "default" | "inherit" | "large" | "medium" | "small";
}

export default Checkbox;
