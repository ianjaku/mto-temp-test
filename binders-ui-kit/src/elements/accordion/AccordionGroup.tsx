import * as React from "react";
import Accordion from "./Accordion";

export interface IAccordionGroupProps {
    children?: React.ReactElement<Accordion> | Array<React.ReactElement<Accordion>>;
    isMultiOpen?: boolean;
    initialOpenedIndex?: number;
    onChangeOpenIndexes?: (openIndexes) => void;
}

export interface IAccordionGroupState {
    openedIndexes: number[];
}

class AccordionGroup extends React.Component<IAccordionGroupProps, IAccordionGroupState> {

    private static defaultProps = {
        isMultiOpen: false,
    };

    constructor(props: IAccordionGroupProps) {
        super(props);

        this.isAccordionOpenedAt = this.isAccordionOpenedAt.bind(this);
        this.toggleAccordionAt = this.toggleAccordionAt.bind(this);
        this.renderAccordion = this.renderAccordion.bind(this);


        const { initialOpenedIndex } = props;
        this.state = {
            openedIndexes: initialOpenedIndex !== undefined ? [initialOpenedIndex] : [],
        };
    }

    public isAccordionOpenedAt(index: number): boolean {
        return this.state.openedIndexes.indexOf(index) > -1;
    }

    public closeAll(setStateCallback: () => void): void {
        this.setState({
            openedIndexes: [],
        }, setStateCallback);
    }

    public toggleAccordionAt(index: number): void {
        const { isMultiOpen, onChangeOpenIndexes } = this.props;
        const { openedIndexes } = this.state;
        const isIndexOpen = openedIndexes.indexOf(index) > -1;
        let newOpenedIndexes;
        if (!isMultiOpen) {
            newOpenedIndexes = isIndexOpen ? [] : [index];
        } else {
            newOpenedIndexes = isIndexOpen ?
                openedIndexes.filter(value => value !== index) :
                openedIndexes.concat([index]);
        }
        this.setState({ openedIndexes: newOpenedIndexes });
        if (onChangeOpenIndexes) {
            onChangeOpenIndexes([index]);
        }
    }

    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    public renderAccordion(accordion, index: number) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        return React.cloneElement(accordion as React.ReactElement<any>, {
            isOpened: this.isAccordionOpenedAt(index),
            onToggle: () => this.toggleAccordionAt(index),
        });
    }

    public render(): JSX.Element {
        const { children } = this.props;

        return (
            <div className="accordion-group">
                {React.Children.map(children, this.renderAccordion)}
            </div>
        );
    }
}

export default AccordionGroup;
