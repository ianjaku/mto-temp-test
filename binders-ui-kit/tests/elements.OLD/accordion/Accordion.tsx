import * as React from "react";
import { Accordion, AccordionGroup } from "../../../src/elements/accordion";
import { create } from "react-test-renderer";
import { mount } from "enzyme";
import wrapWithTheme from "../../themeHelper";

const DEFAULT_HEADER = "Default Header";

const makeAccordionHeader = (content: string) => <div>{content}</div>;

const createAccordion = (isOpened = false) => (
    wrapWithTheme(<Accordion isOpened={isOpened} header={makeAccordionHeader(DEFAULT_HEADER)}>
        <div>{DEFAULT_HEADER}</div>
    </Accordion>)
);

const createAccordionGroup = (isMultiOpen = false, initialOpenedIndex: number = undefined) => (
    wrapWithTheme(<AccordionGroup isMultiOpen={isMultiOpen} initialOpenedIndex={initialOpenedIndex}>
        <Accordion header={makeAccordionHeader("Header #1")}>
            <div>Content #1</div>
        </Accordion>
        <Accordion header={makeAccordionHeader("Header #2")}>
            <div>Content #2</div>
        </Accordion>
        <Accordion header={makeAccordionHeader("Header #3")}>
            <div>Content #3</div>
        </Accordion>
    </AccordionGroup>)
);

describe("Accordion", () => {
    test("Closed Accordion (snapshot)", () => {
        const accordion = create(createAccordion());
        const serialized = accordion.toJSON();
        expect(serialized).toMatchSnapshot();
    });

    test("Opened Accordion (snapshot)", () => {
        const accordion = create(createAccordion(true));
        const serialized = accordion.toJSON();
        expect(serialized).toMatchSnapshot();
    });

    test("Accordion Group (snapshot)", () => {
        const accordion = create(createAccordionGroup());
        const serialized = accordion.toJSON();
        expect(serialized).toMatchSnapshot();
    });

    test("Closed Accordion children visibility (shallow: enzyme+jsdom)", () => {
        const accordion = mount(createAccordion());
        const accordionBody = accordion.find(".accordion-body").exists();
        expect(accordionBody).toBe(false);
    });

    test("Opened Accordion children visibility (shallow: enzyme+jsdom)", () => {
        const visible = true;
        const accordion = mount(createAccordion(visible));
        const accordionBody = accordion.find(".accordion-body").exists();
        expect(accordionBody).toBe(true);
    });

    test("All closed Accordion Group (shallow: enzyme+jsdom)", () => {
        const accordion = mount(createAccordionGroup());
        const accordionHeaders = accordion.find(".accordion-header");
        const accordionBodies = accordion.find(".according-body");
        expect(accordionHeaders).toHaveLength(3);
        expect(accordionBodies).toHaveLength(0);
    });

    test("One Accordion from a Group should be open if initialOpenedIndex is set (mount: enzyme+jsdom)", () => {
        const isMultiOpen = false;
        const initialOpenedIndex = 1;
        const accordion = mount(createAccordionGroup(isMultiOpen, initialOpenedIndex));
        const accordionHeaders = accordion.find(".accordion-header");
        const accordionBodies = accordion.find(".accordion-body");
        expect(accordionHeaders).toHaveLength(3);
        expect(accordionBodies).toHaveLength(1);
    });

    test("One Accordion from a Group should open when a header is clicked and then closed (mount: enzyme+jsdom)", () => {
        const accordion = mount(createAccordionGroup());
        const accordionHeaders = accordion.find(".accordion-header");
        const accordionBodies = accordion.find(".accordion-body");
        expect(accordionHeaders).toHaveLength(3);
        expect(accordionBodies).toHaveLength(0);
        const firstAccordionHeader = accordionHeaders.get(0);
        firstAccordionHeader.props.onClick({ target: null });
        accordion.update();
        expect(accordion.find(".accordion-body")).toHaveLength(1);
        firstAccordionHeader.props.onClick({ target: null });
        accordion.update();
        expect(accordion.find(".accordion-body")).toHaveLength(0);
    });

    test("Multi open accordions should open and close correctly (shallow: enzyme+jsdom)", () => {
        const accordion = mount(createAccordionGroup(true, undefined));
        const accordionHeaders = accordion.find(".accordion-header");
        const accordionBodies = accordion.find(".accordion-body");
        expect(accordionHeaders).toHaveLength(3);
        expect(accordionBodies).toHaveLength(0);
        const firstAccordionHeader = accordionHeaders.get(0);
        firstAccordionHeader.props.onClick({ target: null });
        accordion.update();
        expect(accordion.find(".accordion-body")).toHaveLength(1);
        const secondAccordionHeader = accordionHeaders.get(1);
        secondAccordionHeader.props.onClick({ target: null });
        accordion.update();
        expect(accordion.find(".accordion-body")).toHaveLength(2);
        secondAccordionHeader.props.onClick({ target: null });
        accordion.update();
        expect(accordion.find(".accordion-body")).toHaveLength(1);
    });
});
