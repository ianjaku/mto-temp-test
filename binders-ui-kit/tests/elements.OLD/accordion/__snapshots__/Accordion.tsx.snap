// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion Accordion Group (snapshot) 1`] = `
<div
  className="accordion-group"
>
  <div
    className="accordion"
  >
    <div
      className="accordion-header"
      onClick={[Function]}
    >
      <span
        className="material-icons accordion-arrow"
        color=""
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "display": "inline-block",
            "fontSize": 24,
            "muiPrepared": true,
            "position": "relative",
            "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
            "userSelect": "none",
          }
        }
      >
        arrow_drop_down
      </span>
      <div>
        Header #1
      </div>
    </div>
  </div>
  <div
    className="accordion"
  >
    <div
      className="accordion-header"
      onClick={[Function]}
    >
      <span
        className="material-icons accordion-arrow"
        color=""
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "display": "inline-block",
            "fontSize": 24,
            "muiPrepared": true,
            "position": "relative",
            "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
            "userSelect": "none",
          }
        }
      >
        arrow_drop_down
      </span>
      <div>
        Header #2
      </div>
    </div>
  </div>
  <div
    className="accordion"
  >
    <div
      className="accordion-header"
      onClick={[Function]}
    >
      <span
        className="material-icons accordion-arrow"
        color=""
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        style={
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "display": "inline-block",
            "fontSize": 24,
            "muiPrepared": true,
            "position": "relative",
            "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
            "userSelect": "none",
          }
        }
      >
        arrow_drop_down
      </span>
      <div>
        Header #3
      </div>
    </div>
  </div>
</div>
`;

exports[`Accordion Closed Accordion (snapshot) 1`] = `
<div
  className="accordion"
>
  <div
    className="accordion-header"
    onClick={[Function]}
  >
    <span
      className="material-icons accordion-arrow"
      color=""
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={
        Object {
          "color": "rgba(0, 0, 0, 0.87)",
          "display": "inline-block",
          "fontSize": 24,
          "muiPrepared": true,
          "position": "relative",
          "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
          "userSelect": "none",
        }
      }
    >
      arrow_drop_down
    </span>
    <div>
      Default Header
    </div>
  </div>
</div>
`;

exports[`Accordion Opened Accordion (snapshot) 1`] = `
<div
  className="accordion is-opened"
>
  <div
    className="accordion-header is-opened"
    onClick={[Function]}
  >
    <span
      className="material-icons accordion-arrow is-opened"
      color=""
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      style={
        Object {
          "color": "rgba(0, 0, 0, 0.87)",
          "display": "inline-block",
          "fontSize": 24,
          "muiPrepared": true,
          "position": "relative",
          "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
          "userSelect": "none",
        }
      }
    >
      arrow_drop_down
    </span>
    <div>
      Default Header
    </div>
  </div>
  <div
    className="accordion-body"
  >
    <div>
      Default Header
    </div>
  </div>
</div>
`;
